File: frontend/src/presenters/flipOpportunityPresenter.test.ts
import { describe, expect, it } from 'vitest';
import {
  classifyMargin,
  formatMargin,
  formatProfit,
  formatQuantity,
  formatVolume,
} from './flipOpportunityPresenter';

describe('classifyMargin', () => {
  // Provisional threshold (docs/TECH_STACK.md notes this is TBD once real
  // data ranges are visible -- the mockup's own four example rows don't
  // encode a consistent margin-based rule).
  it('classifies margins at or above 50% as "up"', () => {
    expect(classifyMargin(97.84)).toBe('up');
    expect(classifyMargin(50)).toBe('up');
  });

  it('classifies margins below 50% as "mid"', () => {
    expect(classifyMargin(49.9)).toBe('mid');
    expect(classifyMargin(7)).toBe('mid');
  });
});

describe('formatMargin', () => {
  it('formats a positive margin with a + sign and rounds to a whole percent', () => {
    expect(formatMargin(97.84)).toEqual({ text: '+98%', colorClass: 'up' });
  });

  it('formats a small margin as "mid"', () => {
    expect(formatMargin(7.2)).toEqual({ text: '+7%', colorClass: 'mid' });
  });

  it('formats a zero margin without a + sign issue', () => {
    expect(formatMargin(0)).toEqual({ text: '+0%', colorClass: 'mid' });
  });
});

describe('formatProfit', () => {
  it('formats profit to 2 decimal places with a + sign, colored by margin', () => {
    expect(formatProfit(0.9784, 97.84)).toEqual({ text: '+0.98', colorClass: 'up' });
  });

  it('colors profit "mid" when the margin is below the threshold, regardless of profit size', () => {
    expect(formatProfit(14, 7)).toEqual({ text: '+14.00', colorClass: 'mid' });
  });
});

describe('formatVolume', () => {
  it('abbreviates thousands as "k" with one decimal, trimming a trailing .0', () => {
    expect(formatVolume(1234)).toBe('1.2k');
    expect(formatVolume(2000)).toBe('2k');
  });

  it('abbreviates millions as "M"', () => {
    expect(formatVolume(1_500_000)).toBe('1.5M');
  });

  it('leaves sub-1000 volumes as a plain rounded number', () => {
    expect(formatVolume(999)).toBe('999');
    expect(formatVolume(54)).toBe('54');
  });
});

describe('formatQuantity', () => {
  it('shows exact integers without the ≈ prefix', () => {
    expect(formatQuantity(1)).toBe('1');
    expect(formatQuantity(366)).toBe('366');
  });

  it('shows non-integers with the ≈ prefix, rounded to 2 decimals, trimming trailing zeros', () => {
    expect(formatQuantity(1.9784)).toBe('≈1.98');
    expect(formatQuantity(1.999)).toBe('≈2');
  });
});

File: app/test/poe_flip_finder/flip_opportunity_presenter_test.exs
defmodule PoeFlipFinder.FlipOpportunityPresenterTest do
  use ExUnit.Case, async: true

  alias PoeFlipFinder.FlipOpportunityPresenter

  # Ported 1:1 from flipOpportunityPresenter.test.ts.

  describe "classify_margin/1" do
    # Provisional threshold (docs/TECH_STACK.md notes this is TBD once real
    # data ranges are visible -- the mockup's own four example rows don't
    # encode a consistent margin-based rule).
    test "classifies margins at or above 50% as :up" do
      assert FlipOpportunityPresenter.classify_margin(97.84) == :up
      assert FlipOpportunityPresenter.classify_margin(50) == :up
    end

    test "classifies margins below 50% as :mid" do
      assert FlipOpportunityPresenter.classify_margin(49.9) == :mid
      assert FlipOpportunityPresenter.classify_margin(7) == :mid
    end
  end

  describe "format_margin/1" do
    test "formats a positive margin with a + sign and rounds to a whole percent" do
      assert FlipOpportunityPresenter.format_margin(97.84) == %{text: "+98%", color_class: :up}
    end

    test "formats a small margin as :mid" do
      assert FlipOpportunityPresenter.format_margin(7.2) == %{text: "+7%", color_class: :mid}
    end

    test "formats a zero margin without a sign issue" do
      assert FlipOpportunityPresenter.format_margin(0) == %{text: "+0%", color_class: :mid}
    end
  end

  describe "format_profit/2" do
    test "formats profit to 2 decimal places with a + sign, colored by margin" do
      assert FlipOpportunityPresenter.format_profit(0.9784, 97.84) == %{
               text: "+0.98",
               color_class: :up
             }
    end

    test "colors profit :mid when the margin is below the threshold, regardless of profit size" do
      assert FlipOpportunityPresenter.format_profit(14, 7) == %{text: "+14.00", color_class: :mid}
    end
  end

  describe "format_volume/1" do
    test "abbreviates thousands as k with one decimal, trimming a trailing .0" do
      assert FlipOpportunityPresenter.format_volume(1234) == "1.2k"
      assert FlipOpportunityPresenter.format_volume(2000) == "2k"
    end

    test "abbreviates millions as M" do
      assert FlipOpportunityPresenter.format_volume(1_500_000) == "1.5M"
    end

    test "leaves sub-1000 volumes as a plain rounded number" do
      assert FlipOpportunityPresenter.format_volume(999) == "999"
      assert FlipOpportunityPresenter.format_volume(54) == "54"
    end
  end

  describe "format_quantity/1" do
    test "shows exact integers without the ~ prefix" do
      assert FlipOpportunityPresenter.format_quantity(1) == "1"
      assert FlipOpportunityPresenter.format_quantity(366) == "366"
    end

    test "shows non-integers with the ~ prefix, rounded to 2 decimals, trimming trailing zeros" do
      assert FlipOpportunityPresenter.format_quantity(1.9784) == "≈1.98"
      assert FlipOpportunityPresenter.format_quantity(1.999) == "≈2"
    end
  end
end
